controller/src/controller.rs

```diff
@@ -247,25 +253,54 @@ impl<T: BottlerocketShadowClient> BrupopController<T> {
             let active_set_size = active_set.len();
             event!(Level::TRACE, ?active_set, "Found active set of nodes.");
 
-            if !active_set.is_empty() {
-                // Try to push forward all active nodes, gathering results along the way.
-                let mut active_brss: Vec<BottlerocketShadow> = active_set.into_values().collect();
-
-                for brs in active_brss.drain(..) {
-                    // Timeouts and errors are logged by instrumentation in `progress_node()`.
-                    #[allow(unused_must_use)]
-                    {
-                        self.progress_node(brs).await;
+            // update time window: users can specify a update time window to operate Bottlerocket nodes update. If current time isn't within the time window,
+            // controller shouldn't have any action on it.
+            match within_time_window()? {
+                // If controller has already started a node update but the update time window stops,
+                // controller should respect it, continue to complete the update, and stop actions on remaining nodes.
+                // This logic will cooperate with 6 mins buffer strategy.
+                false => {
+                    // try to find out if any nodes are being updated. If yes, controller
+                    // will complete them and pause actions on other waitingForUpdate nodes.
+                    if !active_set.is_empty() {
+                        // Try to push forward all active nodes, gathering results along the way.
+                        let mut active_brss: Vec<BottlerocketShadow> =
+                            active_set.into_values().collect();
+
+                        for brs in active_brss.drain(..) {
+                            // Timeouts and errors are logged by instrumentation in `progress_node()`.
+                            #[allow(unused_must_use)]
+                            {
+                                self.progress_node(brs).await;
+                            }
+                        }
+                    } else {
+                        sleep(ACTION_INTERVAL).await;
+                        continue;
                     }
                 }
-            }
+                true => {
+                    if !active_set.is_empty() {
+                        // Try to push forward all active nodes, gathering results along the way.
+                        let mut nodes: Vec<BottlerocketShadow> = active_set.into_values().collect();
+
+                        for brs in nodes.drain(..) {
+                            // Timeouts and errors are logged by instrumentation in `progress_node()`.
+                            #[allow(unused_must_use)]
+                            {
+                                self.progress_node(brs).await;
+                            }
+                        }
+                    }
```

Comment:
I see this chunk is pretty similar to the `false` block above. Would it make sense to refactor this into it's own function since they are so similar?

Reply:
Also, I think I like the variable name `nodes` better than `brss` which I'm guessing stands for "bottle rocket shadows"?